File: Third semester/Synthesis and analysis of algorithms/Exercises/LinkedList/LinkedList/LinkedList.cs
using System;

namespace LinkedList
{
    public class LinkedList<T>
    {
        private Node<T> head;

        public void InsertFront(T newData)
        {
            Node<T> newNode = new Node<T>(newData);
            newNode.next = head;
            head = newNode;
        }

        public void InsertLast(T newData)
        {
            Node<T> newNode = new Node<T>(newData);

            if (head == null)
            {
                head = newNode;
                return;
            }

            Node<T> lastNode = GetLastNode();
            lastNode.next = newNode;
        }

        public void InsertAfter(Node<T> previousNode, T newData)
        {
            if (previousNode == null)
            {
                throw new InvalidNodeValueException();
            }

            Node<T> newNode = new Node<T>(newData);
            newNode.next = previousNode.next;
            previousNode.next = newNode;
        }

        public void PrintNodes()
        {
            Node<T> current = head;

            while (current != null)
            {
                Console.WriteLine(current.data);
                current = current.next;
            }
        }

        private Node<T> GetLastNode()
        {
            Node<T> temp = head;

            while (temp.next != null)
            {
                temp = temp.next;
            }

            return temp;
        }
    }
}

File: Third semester/Java/Exercises/LinkedList/src/LinkedList.java
import java.util.Iterator;

public class LinkedList<T> implements Iterable<T> {
    private Node head;

    public void insertFront(Integer newData) {
        Node newNode = new Node(newData);
        newNode.next = head;
        head = newNode;
    }

    public void insertLast(Integer newData) {
        Node newNode = new Node(newData);

        if (head == null) {
            head = newNode;
            return;
        }

        Node lastNode = getLastNode();
        lastNode.next = newNode;
    }

    public void printList() {
        Node current = head;

        while (current != null) {
            System.out.println(current.data);
            current = current.next;
        }
    }

    private Node getLastNode() {
        Node temp = head;

        while (temp.next != null) {
            temp = temp.next;
        }

        return temp;
    }

    public int size() {
        int size = 0;
        Node temp = head;

        while (temp.next != null) {
            size++;
            temp = temp.next;
        }

        return size + 1;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            Node current = head;

            @Override
            public boolean hasNext() {
                return current != null;
            }

            @Override
            public T next() {
                if (hasNext()) {
                    T data = (T)current.data;
                    current = current.next;
                    return data;
                }

                return null;
            }
        };
    }
}
